iree/hal/local/loaders/wasm_module_loader.c

```diff
@@ -0,0 +1,552 @@
+        &iree_hal_wasm_executable_vtable, executable_layout_count,
+        executable_layouts, executable_layouts_ptr, host_allocator,
+        &executable->base);
+    executable->def = executable_def;
+    executable->entry_fn_count = entry_point_count;
+  }
+
+  if (iree_status_is_ok(status)) {
+    status = iree_hal_wasm_executable_load(executable, host_allocator);
+  }
+
+  if (iree_status_is_ok(status)) {
+    status = iree_hal_wasm_executable_resolve_symbols(executable);
+  }
+
+  if (iree_status_is_ok(status)) {
+    *out_executable = (iree_hal_executable_t*)executable;
+  } else {
+    iree_hal_executable_release((iree_hal_executable_t*)executable);
+  }
+
+  IREE_TRACE_ZONE_END(z0);
+  return status;
+}
+
+static void iree_hal_wasm_executable_destroy(
+    iree_hal_executable_t* base_executable) {
+  iree_hal_wasm_executable_t* executable =
+      (iree_hal_wasm_executable_t*)base_executable;
+  iree_allocator_t host_allocator = executable->base.host_allocator;
+  IREE_TRACE_ZONE_BEGIN(z0);
+
+  // TODO(scotttodd): ref counting / sharing between threads?
+  if (executable->wasm_exec_env) {
+    wasm_runtime_destroy_exec_env(executable->wasm_exec_env);
+  }
+  if (executable->wasm_module_inst) {
+    wasm_runtime_deinstantiate(executable->wasm_module_inst);
+  }
+  if (executable->wasm_module) {
+    wasm_runtime_unload(executable->wasm_module);
+  }
+
+  iree_allocator_free(host_allocator, executable);
+
+  IREE_TRACE_ZONE_END(z0);
+}
+
+static iree_status_t iree_hal_wasm_executable_issue_call(
+    iree_hal_local_executable_t* base_executable, iree_host_size_t ordinal,
+    const iree_hal_executable_dispatch_state_v0_t* dispatch_state,
+    const iree_hal_vec3_t* workgroup_id) {
+  iree_hal_wasm_executable_t* executable =
+      (iree_hal_wasm_executable_t*)base_executable;
+
+  // WebAssembly modules do not have access to any memory outside of their
+  // sandboxed runtime unless explicitly allowed through module imports.
+
+  // TODO(#4022): Use a smarter allocator / allocation scheme.
+  //   This is just MVP placeholder code.
+
+  // Deep clone dispatch_state and workgroup_id using WASM-accessible memory.
+  // All pointers in these structs should be WASM virtual memory addresses, not
+  // the typical host pointers or memory addresses.
+
+  // Clone dispatch_state.
+  void* native_buffer_dispatch_state = NULL;
+  uint32_t wasm_buffer_dispatch_state = wasm_runtime_module_malloc(
+      executable->wasm_module_inst,
+      sizeof(iree_hal_executable_dispatch_state_v0_t),
+      &native_buffer_dispatch_state);
+  if (IREE_UNLIKELY(wasm_buffer_dispatch_state == 0)) {
+    return iree_make_status(IREE_STATUS_RESOURCE_EXHAUSTED, "malloc failed");
+  }
+  iree_hal_executable_dispatch_state_v0_t* wasm_dispatch_state =
+      (iree_hal_executable_dispatch_state_v0_t*)native_buffer_dispatch_state;
+  // Workground count/size.
+  wasm_dispatch_state->workgroup_count.x = dispatch_state->workgroup_count.x;
```

Comment:
hmm I think we can make this slightly better now and easier to incrementally improve by adding a helper fn that takes a dispatch state and an iree_allocator_t and clones it - that way it can do a smarter single malloc vs. these individual ones, and the iree_allocator_t could be defined to pull from the wasm stack vs the heap. Something like `iree_hal_executable_dispatch_state_clone(src, allocator, &dst);`

I think "copy the state into this sandbox stack" will be a common thing (it'll be needed with enclaves, and shared memory sandbox processes, etc) so would be nice to get that together now even if just as a function local in here we migrate out later.

It's worth getting right now so that even if we leave it as a malloc here we are only doing one and the change to using the stack/reusing stack slots would be just modifying the iree_allocator_t vs. all this code.

If you haven't seen it, the pattern used elsewhere is to compute the total_size of the struct (`sizeof(struct) + sizeof(push_constants[0]) * push_constant_count + sizeof(bindings[0]) * binding_count`) and then slicing the pointers out of it. That ensures it's all contiguous.